### PR TITLE
rewritten tag injection and reimplemented scenario tag loading

### PR DESCRIPTION
**Pull request by Kantanomo**
tag injection system in the project has been completely rewritten and overhauled, to not depend on any portion of the std library. 

new system is faster and has more flexibility in the way it can be used. It now supports loading from multiple cache files at once. It can inject tags in single player now and can inject into the dedicated server. Which removes the need for writing code into for dedicated servers to work around injected game state objects. 

the scenario_tags_load function from the game has been rewritten into the project as well as several other functions. the loading of debug information from cache files has also been rewritten into these functions. in which allowed removal of the old tag name loading code from TagInterface. 

several other functions related to accessing tags in the project have added as well as some moved out of Tags::/TagInterface and put in a more accessible location.

### CODE REVIEW

File: xlive/Blam/Engine/tag_files/tag_loader/tag_injection_define.h [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#pragma once
+
+#define K_TAG_INJECTION_DEBUG 0
```
Comment (punpckhdq) [inline xlive/Blam/Engine/tag_files/tag_loader/tag_injection_define.h]: dont use k_ prefix for feature or build configuration defines

File: xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,688 @@
...
+			// Get the offset of the next index
+			lazy_fread(this->m_active_map_file_handle, this->m_active_map_cache_header.tag_name_offsets_offset + sizeof(uint32) * (current_index + 1), &next_offset, sizeof(uint32), 1);
+
+			// Current size is calculated using the offsets of the two indexes
+			current_size = next_offset - current_offset;
+
+			// Read the current debug name
+			lazy_fread(this->m_active_map_file_handle, this->m_active_map_cache_header.tag_name_buffer_offset + current_offset, &name_buffer, current_size, 1);
+
+			if(memcmp(tag_name, name_buffer, strlen(tag_name)) == 0)
```
Comment (nukeulater) [inline xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp]: strcmp?

File: xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+#include "geometry/geometry_block.h"
+#include "models/render_model_definitions.h"
+#include "physics/physics_model_definitions.h"
+#include "render/weather_definitions.h"
+#include "units/biped_definitions.h"
+#include "units/vehicle_definitions.h"
+#include "Util/filesys.h"
+
+#define lazy_fread(FILE, OFFSET, OUT, SIZE, COUNT)\
+	fseek(FILE, OFFSET, SEEK_SET);\
```
Comment (nukeulater) [inline xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp]: wonder what would happen if I were to invoke this macro inside a brace-less if statement scope.

File: xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,688 @@
...
+			// Get the offset of the current index
+			lazy_fread(this->m_active_map_file_handle, this->m_active_map_cache_header.tag_name_offsets_offset + sizeof(uint32) * current_index, &current_offset, sizeof(uint32), 1);
+
+			// Current size is calculated using the total size of the buffer and the current offset;
+			current_size = (this->m_active_map_cache_header.tag_name_offsets_offset + (this->m_active_map_cache_header.debug_tag_name_count * sizeof(uint32))) - current_offset;
+
+			// Read the current debug name
+			lazy_fread(this->m_active_map_file_handle, this->m_active_map_cache_header.tag_name_buffer_offset + current_offset, &name_buffer, current_size, 1);
+
+			if (memcmp(tag_name, name_buffer, strlen(tag_name)) == 0)
```
Comment (nukeulater) [inline xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp]: strcmp?

File: xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,713 @@
...
+#include "creatures/creature_definitions.h"
+#include "geometry/geometry_block.h"
+#include "models/render_model_definitions.h"
+#include "physics/physics_model_definitions.h"
+#include "render/weather_definitions.h"
+#include "units/biped_definitions.h"
+#include "units/vehicle_definitions.h"
+#include "Util/filesys.h"
+
+#define lazy_fread(FILE, OFFSET, OUT, SIZE, COUNT)\
```
Comment (nukeulater) [inline xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp]: could be a function instead

File: xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,713 @@
...
+	
+	cache_file_tag_instance temp_instance;
+
+	int32 current_index = 0;
+	int32 current_offset = 0;
+	uint32 next_offset = 0;
+	uint32 current_size = 0;
+	char name_buffer[MAX_PATH] = {};
+
+	while (true)
```
Comment (nukeulater) [inline xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp]: for loop instead?

File: xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+
+		if(entry->is_injected)
+			continue;
+
+		entry->is_injected = true;
+
+		if(entry->type.group == _tag_group_bitmap)
+		{
+			bitmap_group* test = (bitmap_group*)entry->loaded_data->get_data();
+			auto sdfsd = 213123;
```
Comment (nukeulater) [inline xlive/Blam/Engine/tag_files/tag_loader/tag_injection_manager.cpp]: ?

File: xlive/Blam/Engine/tag_files/tag_loader/tag_injection_table.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+{
+	return this->m_entry_count;
+}
+
+s_tag_injecting_table_entry* c_tag_injection_table::init_entry(datum cache_index, e_tag_group type)
+{
+	//tag_group temp;
+	//temp.group = type;
+
+	if (this->m_entry_count + 1 == this->m_table_size)
```
Comment (nukeulater) [inline xlive/Blam/Engine/tag_files/tag_loader/tag_injection_table.cpp]: wouldn't `>=` be better?

File: xlive/Blam/Engine/tag_files/tag_loader/xml/xml_definition_loader.cpp [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+
+#include "bitmaps/bitmap_group.h"
+#include "cache/cache_files.h"
+#include "tag_files/data_reference.h"
+#include "tag_files/string_id.h"
+#include "tag_files/tag_block.h"
+#include "tag_files/tag_reference.h"
+#include "tag_files/tag_loader/tag_injection.h"
+
+#define lazy_fread(FILE, OFFSET, OUT, SIZE, COUNT)\
```
Comment (nukeulater) [inline xlive/Blam/Engine/tag_files/tag_loader/xml/xml_definition_loader.cpp]: same here, function?

File: xlive/Blam/Engine/tag_files/tag_loader/xml/xml_definition_loader.cpp [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+	{ \
+		fseek(FILE, OFFSET, SEEK_SET);\
+		fread(OUT, SIZE, COUNT, FILE); \
+	}
+
+#define lazy_malloc_buffer(TYPE, COUNT)\
+	(TYPE*)malloc(sizeof(TYPE) * (COUNT))
```
Comment (nukeulater) [inline xlive/Blam/Engine/tag_files/tag_loader/xml/xml_definition_loader.cpp:16]: xd `(COUNT)`

Comment (Kantanomo) [reply xlive/Blam/Engine/tag_files/tag_loader/xml/xml_definition_loader.cpp:16]: xd

**Pull request merged**